trailmet/algorithms/binarize/utils.py

```diff
@@ -0,0 +1,154 @@
+    self.num_classes = num_classes
+    self.epsilon = epsilon
+    self.logsoftmax = nn.LogSoftmax(dim=1)
+
+  def forward(self, inputs, targets):
+    log_probs = self.logsoftmax(inputs)
+    targets = torch.zeros_like(log_probs).scatter_(1, targets.unsqueeze(1), 1)
+    targets = (1 - self.epsilon) * targets + self.epsilon / self.num_classes
+    loss = (-targets * log_probs).mean(0).sum()
+    return loss
+
+class AverageMeter(object):
+    """Computes and stores the average and current value"""
+    def __init__(self, name, fmt=':f'):
+        self.name = name
+        self.fmt = fmt
+        self.reset()
+
+    def reset(self):
+        self.val = 0
+        self.avg = 0
+        self.sum = 0
+        self.count = 0
+
+    def update(self, val, n=1):
+        self.val = val
+        self.sum += val * n
+        self.count += n
+        self.avg = self.sum / self.count
+
+    def __str__(self):
+        fmtstr = '{name} {val' + self.fmt + '} ({avg' + self.fmt + '})'
+        return fmtstr.format(**self.__dict__)
+    
+class ProgressMeter(object):
+    def __init__(self, num_batches, meters, prefix=""):
+        self.batch_fmtstr = self._get_batch_fmtstr(num_batches)
+        self.meters = meters
+        self.prefix = prefix
+
+    def display(self, batch):
+        entries = [self.prefix + self.batch_fmtstr.format(batch)]
+        entries += [str(meter) for meter in self.meters]
+        print('\t'.join(entries))
+
+    def _get_batch_fmtstr(self, num_batches):
+        num_digits = len(str(num_batches // 1))
+        fmt = '{:' + str(num_digits) + 'd}'
+        return '[' + fmt + '/' + fmt.format(num_batches) + ']'
+
+def save_checkpoint(state, is_best, save, dataset):
+    if not os.path.exists(save):
+        os.makedirs(save)
+    if dataset=='cifar100':
+        filename = os.path.join(save, 'cifar100-checkpoint.pth.tar')
+        torch.save(state, filename)
+        if is_best:
+            best_filename = os.path.join(save, 'cifar100-model_best.pth.tar')
+            shutil.copyfile(filename, best_filename)
+    if dataset=='cifar10':
+        filename = os.path.join(save, 'cifar10-checkpoint.pth.tar')
+        torch.save(state, filename)
+        if is_best:
+            best_filename = os.path.join(save, 'cifar10-model_best.pth.tar')
+            shutil.copyfile(filename, best_filename)
+    if dataset=='tiny-imagenet':
+        filename = os.path.join(save, 'tiny-imagenet-checkpoint.pth.tar')
+        torch.save(state, filename)
+        if is_best:
+            best_filename = os.path.join(save, 'tiny-imagenet-model_best.pth.tar')
+            shutil.copyfile(filename, best_filename)
+
+def adjust_learning_rate(optimizer, epoch, args):
+    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
+    lr = args.lr * (0.1 ** (epoch // 30))
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = lr
+        
+def accuracy(output, target, topk1=(1,)):
```

Comment:
Can this be a part of trailmet utils?

Reply:
should i make the save_checkpoint function generalized or is the implementation for the 3 datasets enough?